File: backend/internal/model/certificate.go
/*
 * @Author              : 寂情啊
 * @Date                : 2025-12-23 13:54:36
 * @LastEditors         : 寂情啊
 * @LastEditTime        : 2025-12-23 13:54:50
 * @FilePath            : frp-web-testbackendinternalmodelcertificate.go
 * @Description         : 说明
 * 倾尽绿蚁花尽开，问潭底剑仙安在哉
 */
package model

import (
	"time"
)

// 证书状态常量
const (
	CertStatusPending  = "pending"  // 申请中
	CertStatusActive   = "active"   // 有效
	CertStatusExpiring = "expiring" // 即将过期
	CertStatusExpired  = "expired"  // 已过期
	CertStatusFailed   = "failed"   // 申请失败
)

// Certificate SSL证书模型
type Certificate struct {
	ID            uint       `json:"id" gorm:"primaryKey"`
	ProxyID       uint       `json:"proxy_id" gorm:"index"`                 // 关联的代理ID
	Domain        string     `json:"domain" gorm:"size:255;not null"`       // 域名
	ProviderID    uint       `json:"provider_id" gorm:"index"`              // DNS提供商ID
	Status        string     `json:"status" gorm:"size:20;default:pending"` // 状态
	CertPEM       string     `json:"cert_pem" gorm:"type:text"`             // 证书内容(PEM格式)
	KeyPEM        string     `json:"-" gorm:"type:text"`                    // 私钥内容(PEM格式，不返回给前端)
	IssuerCertPEM string     `json:"issuer_cert_pem" gorm:"type:text"`      // 颁发者证书
	NotBefore     *time.Time `json:"not_before"`                            // 生效时间
	NotAfter      *time.Time `json:"not_after"`                             // 过期时间
	LastError     string     `json:"last_error" gorm:"type:text"`           // 最后错误信息
	AutoRenew     bool       `json:"auto_renew" gorm:"default:true"`        // 是否自动续期
	AcmeAccountID string     `json:"acme_account_id" gorm:"size:255"`       // ACME账户ID
	CreatedAt     time.Time  `json:"created_at"`
	UpdatedAt     time.Time  `json:"updated_at"`
}

// IsExpiringSoon 检查证书是否即将过期（30天内）
func (c *Certificate) IsExpiringSoon() bool {
	if c.NotAfter == nil {
		return false
	}
	return time.Until(*c.NotAfter) < 30*24*time.Hour
}

// IsExpired 检查证书是否已过期
func (c *Certificate) IsExpired() bool {
	if c.NotAfter == nil {
		return false
	}
	return time.Now().After(*c.NotAfter)
}

File: web/src/api/certificate.ts
import request from './request';

export interface Certificate {
  id: number;
  proxy_id: number;
  domain: string;
  provider_id: number;
  status: 'pending' | 'active' | 'expiring' | 'expired' | 'failed';
  cert_pem: string;
  issuer_cert_pem: string;
  not_before: string | null;
  not_after: string | null;
  last_error: string;
  auto_renew: boolean;
  acme_account_id: string;
  created_at: string;
  updated_at: string;
}

export interface RequestCertificateInput {
  proxy_id: number;
  domain: string;
  dns_provider_id: number;
  auto_renew?: boolean;
}

export interface CertificateDownload {
  domain: string;
  cert_pem: string;
  issuer_cert_pem: string;
  full_chain_pem: string;
}

export const certificateApi = {
  list: () => request.get<Certificate[]>('/certificates'),
  getById: (id: number) => request.get<Certificate>(`/certificates/${id}`),
  request: (data: RequestCertificateInput) => request.post<Certificate>('/certificates', data),
  renew: (id: number) => request.post(`/certificates/${id}/renew`),
  reapply: (id: number) => request.post(`/certificates/${id}/reapply`),
  updateAutoRenew: (id: number, autoRenew: boolean) => request.put(`/certificates/${id}/auto-renew`, { auto_renew: autoRenew }),
  download: (id: number) => request.get<CertificateDownload>(`/certificates/${id}/download`),
  delete: (id: number) => request.delete(`/certificates/${id}`),
  getByDomain: (domain: string) => request.get<Certificate[]>('/certificates/by-domain', { params: { domain } }),
  getExpiring: () => request.get<Certificate[]>('/certificates/expiring'),
  getActive: () => request.get<Certificate[]>('/certificates/active'),
  getMatching: (domain: string) => request.get<Certificate[]>('/certificates/match', { params: { domain } }),
};